### ISSUE TITLE
Transforms revamp

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/boost-histogram#211]
Here's my proposal for transforms:

1. We add `CTransform(cpp_forward, cpp_reverse)`, and eventually a pure python version, that take a compiled function pair and Python function piar, respectively. (I've not done this in Pybind11 before, but it looks pretty easy, and you have experience with it in iMinuit that I might call on ;) )
2. We use this to implement `log = CTransform(_log_fn, _exp_fn)`, where we compile these functions internally.
3. We remove `Sqrt()`. We set `sqrt = Pow(0.5)`.

Now, we have no name duplication based on capital letters, we have clear distinctions, and we have more functionality, and simpler C++ axes.

Let's *try* to get this into 0.6.0, since it will be a breaking change to serialization. Not worth holding up the release too long for, though.

**Comment by HDembinski** [comment]
This is an excellent proposal and may allows us to cut down on the code generated for regular axis. We should definitely implement this.

Before we replace specializations of regular axis, however, we should make a benchmark, to see whether the fill speed is not reduced too much. I expect a noticeable a performance reduction, because even when the CTransform is used, there is an additional function call (and associated pointer jump) needed for every single value that is passed to the axis. Since the number of CPU cycles spend by the regular axis itself is very small, this should show up in the benchmarks. 

Basically we should weigh the reduction in binary size against the reduction in fill performance, and then make a decision of . If there is a noticable reduction in binary size at a modest cost in fill performance, I am all for using the CTransform instead of some of the boost::histogram transforms.

In any case, I am in favor of dropping the explicit `regular_sqrt` template instantiation and express that through the more general `regular_pow`. This may be a bit slower, but probably not by much.

**Comment by HDembinski** [comment]
Thinking more about it, the PyTransform is going to have horrible performance, because you need to make a Python call for each value, even if you use `h.fill(...)`. The PyTransform does not really fit well into the boost::histogram infrastructure. What we need is an axis that accepts a chunk of values and returns a chunk of indices, to amortize the cost of calling a Python transform. This requires more thinking on how to implement this. Please let us do this after the 1.0 release.

**Comment by HDembinski** [comment]
The CTransform we should try out now, though.

**Comment by HDembinski** [comment]
Remember that you need to pass two function pointers, one for the forward transform and one for the backward transform.

**Comment by henryiii** [comment]
Okay, let's go forward with the CTransform and see how that performs. If we are even close to the speed of Log (factor 2-3 or maybe even more), then I think it would be fine to replace it. It should still scale like a Regular axis and not like a Variable one.

**Comment by HDembinski** [comment]
Yes, it is O(1) but with a larger factor in front. A factor of two reduction in filling speed may increase the time for data processing from 5 min to 10 min.

**Comment by henryiii** [comment]
Currently there is no way to even do this in Python without going to variable bins, so I think we are already a mile ahead. We can do optimizations later, but we should probably start light and flexible. Now a factor of 10 or so would probably change my opinion. But *probably* not 2, not on a specialty transform, not on something that is quite speedy anyway (Regular axes).
